tools/PsesLogAnalyzer/Parse-PsesLog.ps1

```diff
@@ -0,0 +1,268 @@
+        }
+
+        $filestream =
+            [System.IO.FileStream]::new(
+                $Path,
+                [System.IO.FileMode]:: Open,
+                [System.IO.FileAccess]::Read,
+                [System.IO.FileShare]::ReadWrite,
+                4096,
+                [System.IO.FileOptions]::SequentialScan)
+
+        $streamReader = [System.IO.StreamReader]::new($filestream, [System.Text.Encoding]::UTF8)
+
+        function nextLine() {
+            if ($null -ne $peekBuf) {
+                $line = $peekBuf
+                $script:peekBuf = $null
+            }
+            else {
+                $line = $streamReader.ReadLine()
+            }
+
+            $script:currentLineNum++
+            $line
+        }
+
+        function peekLine() {
+            if ($null -ne $peekBuf) {
+                $line = $peekBuf;
+            }
+            else {
+                $line = $script:peekBuf = $streamReader.ReadLine()
+            }
+
+            $line
+        }
+
+        function parseLogEntryStart([string]$line) {
+            if ($DebugTimingInfo) {
+                $sw = [System.Diagnostics.Stopwatch]::StartNew()
+            }
+
+            while ($line -notmatch $logEntryRegex) {
+                Write-Warning "Ignoring line:${currentLineNum} '$line'"
+                $line = nextLine
+            }
+
+            if (!$HideProgress -and ($script:logEntryIndex % 50 -eq 0)) {
+                Write-Progress "Processing log entry ${script:logEntryIndex} on line: ${script:currentLineNum}"
+            }
+
+            [string]$timestampStr = $matches["ts"]
+            [DateTime]$timestamp = $timestampStr
+            [PsesLogLevel]$logLevel = $matches["lev"]
+            [int]$threadId = $matches["tid"]
+            [string]$method = $matches["meth"]
+            [string]$file = $matches["file"]
+            [int]$lineNumber = $matches["line"]
+
+            $message = parseLogMessage $method
+
+            [PsesLogEntry]::new($script:logEntryIndex, $timestamp, $timestampStr, $logLevel, $threadId, $method,
+                $file, $lineNumber, $message.LogMessageType, $message.LogMessage)
+
+            if ($DebugTimingInfo) {
+                $sw.Stop()
+                if ($sw.ElapsedMilliseconds -gt $DebugTimingThresholdMs) {
+                    Write-Warning "Time to parse log entry ${script:logEntryIndex} - $($sw.ElapsedMilliseconds) ms"
+                }
+            }
+
+            $script:logEntryIndex++
+        }
+
+        function parseLogMessage([string]$Method) {
+            $result = [PSCustomObject]@{
+                LogMessageType = [PsesLogMessageType]::Log
+                LogMessage = $null
+            }
+
+            $line = nextLine
+            if ($null -eq $line) {
+                Write-Warning "$($MyInvocation.MyCommand.Name) encountered end of file early."
+                return $result
+            }
+
+            if (($Method -eq 'ReadMessage') -and
```

Comment:
Given a cascade of `if`/`elseif`/`else` statements followed by a single implicit return, this might be refactorable as a series of `if (...) { return }` statements for slightly lower complexity.

Reply:
This is a bigger change. I'll adress it later (in another PR) when I'm back from New York.